src/solver/sgd/forward_sgd.cpp

```diff
@@ -0,0 +1,165 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#include <miopen/conv_solution.hpp>
+#include <miopen/execution_context.hpp>
+#include <miopen/invoke_params.hpp>
+#include <miopen/tensor_view_utils.hpp>
+#include <miopen/sgd/solvers.hpp>
+
+#include <miopen/sgd/invoke_params.hpp>
+#include <miopen/datatype.hpp>
+#include <miopen/mlo_internal.hpp>
+#include <miopen/sgd.hpp>
+#include <miopen/target_properties.hpp>
+
+#define LOCAL_SIZE 256
+
+namespace miopen {
+namespace solver {
+namespace SGD {
+
+bool SGDForward::IsApplicable([[maybe_unused]] const ExecutionContext& constext,
+                              const miopen::SGD::ProblemDescription& problem) const
+{
+    if(!(problem.IsSameLength() && problem.IsSameType() && problem.IsValidLength()))
+    {
+        return false;
+    }
+    if(!(problem.GetParamInDesc().GetType() == miopenFloat ||
+         problem.GetParamInDesc().GetType() == miopenHalf ||
+         problem.GetParamInDesc().GetType() == miopenBFloat16))
+        return false;
+    return true;
+}
+
+ConvSolution SGDForward::GetSolution([[maybe_unused]] const ExecutionContext& context,
+                                     const miopen::SGD::ProblemDescription& problem) const
+{
+    auto result = ConvSolution{miopenStatusSuccess};
+
+    auto dtype        = problem.GetParamInDesc().GetType();
+    auto dims         = problem.GetParamInDesc().GetLengths();
+    auto input_dtype  = miopen::GetDataType(problem.GetParamInDesc().GetType());
+    auto output_dtype = miopen::GetDataType(problem.GetParamOutDesc().GetType());
+
+    bool is_contiguous = problem.IsAllContiguous();
+
+    size_t param_size = std::accumulate(dims.begin(), dims.end(), 1ULL, std::multiplies<size_t>());
+
+    const auto build_params = KernelBuildParameters{
+        {"MIOPEN_USE_FP16", static_cast<int>(dtype == miopenHalf)},
+        {"MIOPEN_USE_FP32", static_cast<int>(dtype == miopenFloat)},
+        {"MIOPEN_USE_FP64", static_cast<int>(dtype == miopenDouble)},
+        {"MIOPEN_USE_BFP16", static_cast<int>(dtype == miopenBFloat16)},
+        {"INPUT_TYPE", input_dtype == "bfloat16" ? "ushort" : input_dtype},
+        {"OUTPUT_TYPE", output_dtype == "bfloat16" ? "ushort" : output_dtype},
```

Comment:
Because of `IsSameType` check, we don't need both.

Reply:
resolved